File: scripts/generate-manifest.js
/**
 * generate-manifest.js
 * Scansiona automaticamente la cartella viaggi/ (incluse tutte le eventuali sottocartelle)
 * e genera manifest.json in modo deterministico e ordinato.
 */

const fs = require('fs');
const path = require('path');

const VIAGGI_DIR = path.join(__dirname, '..', 'viaggi');
const ROOT_DIR = path.join(__dirname, '..');
const OUTPUT_FILE = path.join(__dirname, '..', 'manifest.json');
const VALID_EXTENSIONS = new Set(['.jpg', '.jpeg', '.png', '.webp']);

function naturalCompare(a, b) {
  return a.localeCompare(b, undefined, { numeric: true, sensitivity: 'base' });
}

function scanRecursively(dir) {
  let results = [];
  const entries = fs.readdirSync(dir, { withFileTypes: true });

  for (const entry of entries) {
    const fullPath = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      results = results.concat(scanRecursively(fullPath));
    } else if (entry.isFile()) {
      const ext = path.extname(entry.name).toLowerCase();
      if (VALID_EXTENSIONS.has(ext)) {
        results.push(fullPath);
      }
    }
  }

  return results;
}

function generateManifest() {
  console.log('🔍 Avvio scansione cartella viaggi (con supporto ricorsivo per sottocartelle)...');
  
  if (!fs.existsSync(VIAGGI_DIR)) {
    console.warn('⚠️ Cartella "viaggi" non trovata. Creazione cartella vuota...');
    fs.mkdirSync(VIAGGI_DIR, { recursive: true });
  }

  const entries = fs.readdirSync(VIAGGI_DIR, { withFileTypes: true });
  const manifest = {};
  let totalTrips = 0;
  let totalPhotos = 0;

  // Filtra solo le cartelle dei viaggi principali
  const tripFolders = entries
    .filter(entry => entry.isDirectory())
    .map(entry => entry.name)
    .sort(naturalCompare);

  for (const tripId of tripFolders) {
    const tripPath = path.join(VIAGGI_DIR, tripId);
    
    // Scansiona ricorsivamente (es. viaggi/siviglia/Alcazar/...)
    const fullImagePaths = scanRecursively(tripPath);

    // Ordina i file in modo naturale
    fullImagePaths.sort((a, b) => naturalCompare(a, b));

    // Converti in percorsi relativi compatibili con il web ('/')
    manifest[tripId] = fullImagePaths.map(fullPath => {
      const rel = path.relative(ROOT_DIR, fullPath);
      return rel.split(path.sep).join('/');
    });
    
    totalTrips++;
    totalPhotos += manifest[tripId].length;
    console.log(`📁 Viaggio "${tripId}": ${manifest[tripId].length} fotografie trovate.`);
  }

  // Scrittura del file manifest.json formattato
  fs.writeFileSync(OUTPUT_FILE, JSON.stringify(manifest, null, 2) + '\n', 'utf-8');
  
  console.log(`\n✨ manifest.json generato con successo!`);
  console.log(`📊 Totale viaggi: ${totalTrips} | Totale fotografie: ${totalPhotos}`);
  console.log(`📍 Percorso: ${OUTPUT_FILE}`);
}

generateManifest();

File: scripts/generate-manifest.py
#!/usr/bin/env python3
"""
generate-manifest.py
Scansiona automaticamente la cartella viaggi/ (incluse tutte le eventuali sottocartelle)
e genera manifest.json con percorsi web relativi deterministici.
"""

import os
import json
import re
import sys

if sys.platform == "win32":
    sys.stdout.reconfigure(encoding='utf-8')

VALID_EXTENSIONS = {'.jpg', '.jpeg', '.png', '.webp'}

def natural_key(text):
    return [int(c) if c.isdigit() else c.lower() for c in re.split(r'(\d+)', text)]

def generate_manifest():
    script_dir = os.path.dirname(os.path.abspath(__file__))
    root_dir = os.path.dirname(script_dir)
    viaggi_dir = os.path.join(root_dir, 'viaggi')
    manifest_file = os.path.join(root_dir, 'manifest.json')

    print("[SCAN] Avvio scansione ricorsiva cartella viaggi...")

    if not os.path.exists(viaggi_dir):
        print("[WARN] Cartella 'viaggi' non trovata. Creazione cartella vuota...")
        os.makedirs(viaggi_dir, exist_ok=True)

    manifest = {}
    total_trips = 0
    total_photos = 0

    try:
        entries = sorted(os.listdir(viaggi_dir), key=natural_key)
    except Exception as e:
        print(f"[ERROR] Errore nella lettura di viaggi: {e}")
        return

    for entry in entries:
        trip_path = os.path.join(viaggi_dir, entry)
        if not os.path.isdir(trip_path):
            continue

        trip_id = entry
        images = []

        # Scansione ricorsiva (gestisce sottocartelle come Siviglia/Alcazar, etc.)
        for root, dirs, files in os.walk(trip_path):
            for file in sorted(files, key=natural_key):
                ext = os.path.splitext(file)[1].lower()
                if ext in VALID_EXTENSIONS:
                    full_file_path = os.path.join(root, file)
                    rel_path = os.path.relpath(full_file_path, root_dir)
                    # Converti in slash '/' per il web
                    web_path = rel_path.replace(os.path.sep, '/')
                    images.append(web_path)

        images.sort(key=natural_key)
        manifest[trip_id] = images
        total_trips += 1
        total_photos += len(images)
        print(f"  -> Viaggio '{trip_id}': {len(images)} fotografie trovate.")

    with open(manifest_file, 'w', encoding='utf-8') as f:
        json.dump(manifest, f, indent=2, ensure_ascii=False)
        f.write('\n')

    print(f"\n[OK] manifest.json generato con successo!")
    print(f"[STATS] Totale viaggi: {total_trips} | Totale fotografie: {total_photos}")
    print(f"[DEST] Percorso: {manifest_file}")

if __name__ == '__main__':
    generate_manifest()
